core/src/main/scala/org/apache/spark/CacheManager.scala

```diff
@@ -46,7 +48,14 @@ private[spark] class CacheManager(blockManager: BlockManager) extends Logging {
         val existingMetrics = context.taskMetrics().registerInputMetrics(blockResult.readMethod)
         existingMetrics.incBytesReadInternal(blockResult.bytes)
 
-        val iter = blockResult.data.asInstanceOf[Iterator[T]]
+        val iter = {
+          val dataIter = blockResult.data.asInstanceOf[Iterator[T]]
+          if (blockResult.readMethod != DataReadMethod.Network) {
+            CompletionIterator[T, Iterator[T]](dataIter, blockManager.releaseLock(key))
```

Comment:
Can we bake this logic in blockResult.data and have it return a CompletionIterator? If not, let's move this as a utility in blockResult.